### ISSUE TITLE
Documentation for mixed CNN + LSTM MultiLayerConfiguration for time-series data

### ISSUE THREAD

**Issue by daanvdn** [deeplearning4j/deeplearning4j#5052]
#### Issue Description
I’m looking for some documentation on how to setup a `MultiLayerConfiguration` that combines Convolution and GravesLSTM layers and that takes *variable-length ***time-series data **as input to perform classification (using dl4j v0.9.1). As far as the combo Convolution + LSTM is concerned, I know there is this example: `org.deeplearning4j.examples.multigpu.video.VideoClassificationExample`, but I was wondering if there are any examples outside of the image domain? My data is basically a bunch of clinical indicators in time-series format of varying length. The labels are diagnoses. My feature-set is quite big (between 6k and 10k) so when I try to train a deep neural net with only LSTM layers it takes forever to train just one epoch (like 8hrs / epoch for 1400 minibatches (40gb of data) using 2 p6000 quadro GPUs) . I’ve already looked into using first a few DenseLayers before the LSTMs to reduce the input size for the latter but that degrades the quality a lot. I’ve found a few articles that mention training with CNNs on time-series data, so I wanted to give this approach a try (cf. https://arxiv.org/pdf/1611.06455.pdf). 

I know that obviously the dl4j api supports training CNN classifiers for time-series data through layers like `org.deeplearning4j.nn.conf.layers.Convolution1DLayer, `and`org.deeplearning4j.nn.conf.layers.Subsampling1DLayer` `. `

However, reading the javadoc of these classes it’s unclear to me how I can use these with time-series data of variable length. For instance the `Convolution1DLayer` “treats a multivariate time series with L timesteps and P variables as an L x 1 x P image (L rows high, 1 column wide, P channels deep). The kernel should be H\<L pixels high and W=1 pixels wide.” This seems to presuppose time-series input of fixed length, which I don’t have. Also the `RnnToCnnPreProcessor` and `CnnToRnnPreProcessor` require a fixed  `inputHeight` (i.e. time-series length).  Am I understanding this correctly that variable time-series data can’t be used as is with convolution layers? I assume the easiest way around this is to get the longest time-series length in my data and then ensure that every mini-batch has 1 example with this max length in it and apply padding + masking for the rest of the mini-batch. However, I want to be sure my assumptions are right and that I’m not doing any unnecessary preprocessing.

When it comes to combining several convolution layers with a hidden `org.deeplearning4j.nn.conf.layers.GravesLSTM`, I also can’t figure out how I determine the size of `nIn`. In a configuration that only contains GravesLSTM  layers this is very straightforward as the `nOut` of the previous layer is the `nIn `of the next. However, this does not apply to a mixed CNN+GravesLSTM configuration. Are there any examples that might be of help here? Thanks in advance!

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
int variables = 6000;
int timeSeriesLength = 100;
int nrOfLabels = 5;


MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder().seed(12345)
            .l2(0.001) //l2 regularization on all layers
            .regularization(true)
            .updater(new AdaGrad.Builder().learningRate(0.04).build())
            .list()
            .layer(0, new Convolution1DLayer.Builder(70)
                    //nIn and nOut specify depth. nIn here is the nChannels and nOut is the number of filters to
                    // be applied
                    .nIn(variables).nOut(20).stride(1).activation(Activation.IDENTITY).build())
            .layer(1, new Subsampling1DLayer.Builder(PoolingType.MAX).kernelSize(40).stride(1).build())
            .layer(2, new Convolution1DLayer.Builder(40)
                    .stride(1).nOut(50).activation(Activation.IDENTITY).build())
            .layer(3, new Subsampling1DLayer.Builder(PoolingType.MAX).kernelSize(2).stride(2).build())
            .layer(4, new Convolution1DLayer.Builder(20)
                    .stride(1).nOut(50).activation(Activation.IDENTITY).build())

            .layer(5, new GravesLSTM.Builder()
                    .weightInit(WeightInit.XAVIER)
                    .nIn(/*HOW DO I CALCULATE THIS?*/) <== HOW TO CALCULATE nIn?
                    .nOut(600)
                    .activation(Activation.TANH)
                    .build())
            .layer(6,
                    new RnnOutputLayer.Builder(LossFunctions.LossFunction.XENT).nIn(600)
                            .nOut(nrOfLabels)
                            .activation(Activation.SIGMOID)
                            .build())

            .inputPreProcessor(0, new RnnToCnnPreProcessor(timeSeriesLength, 1, variables))
            .inputPreProcessor(5, new CnnToRnnPreProcessor(timeSeriesLength, 1, 50))
            .backprop(true)
            .pretrain(false)
            .build();
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Version Information

* Deeplearning4j version: using dl4j v0.9.1

**Comment by daanvdn** [comment]
hi @saudet in case you have time and there is no example / documentation available: would it be possible to have a brief look at the `nIn` in layer 5? I have no idea how to calculate this.. thanks in advance

**Comment by AlexDBlack** [comment]
@daanvdn You can try not setting nIn on that layer, and use a ```.setInputType(InputType.recurrent(70))``` instead.
That *should* work and set the nIn automatically for you.

**Comment by daanvdn** [comment]
thanks for the quick reply @AlexDBlack. A final question: is my assumption correct that time series data can only be used with CNNs if all minibatches have the same time series length? (because how would you make the `RnnToCnnPreProcessor` and `CnnToRnnPreProcessor` work?)

**Comment by daanvdn** [comment]
ok I'll give it a go and see what happens, thanks again for the input @AlexDBlack

**Comment by daanvdn** [comment]
@AlexDBlack forgot to ask: when you talk about padding to the size of the longest sequence (+masking), are you referring to the longest sequence within the mini batch or the longest sequence overall? It's this what I'm struggling with in the config above: if it's the longest sequence **overall**, that makes sense to me when creating an `RnnToCnnPreProcessor` or   `CnnToRnnPreProcessor` because they require a fixed `inputHeight` (i.e. time-series length). However, if you are referring to longest sequence with the **mini batch**, then what value should / can be used to create these PreProcessor objects?

**Comment by rickyhai11** [comment]
@daanvdn I have faced to similar issue. Did you solve aforementioned issues?
If yes, could you share your approach?
Thanks

**Comment by liweigu** [comment]
My opinion: masking should be applied to timeSeries' demension, not to miniBatchSize's demension. You may write a LSTM program with masking to figure it out, and then come back to test the LSTM with CNN.
